jupyterhub/apihandlers/servers.py

```diff
@@ -0,0 +1,29 @@
+"""Server handlers"""
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+import json
+
+from .. import orm
+from ..utils import admin_only
+from .base import APIHandler
+
+
+class ServerListAPIHandler(APIHandler):
+    @admin_only
+    def get(self):
+        # TODO(mriedem): Should include_state be conditional on
+        # `spawner.active` like how `user_model` works for
+        # `GET /users`?
+        include_state = True
+        data = [
+            self.server_model(
+                spawner, include_state=include_state, include_user_and_id=True
```

Comment:
I sort of wonder if state and the user-related fields like `url` and `user` should be query parameters which are off by default so that someone calling `GET /servers` that don't care about any of the user information don't need to fetch it from the DB. IOW, I tried to make `GET /servers` similar to how `GET /users` works but maybe that should be opt-in and we should be returning the most minimal representation of each server/spawner by default?

Reply:
While adding the `DELETE /servers/{id}` API in pull #2967 I realized [1] that it's not going to work well in the current form of the API because you need the server id to delete it but only `GET /servers` in this pull request returns the server id and `GET /servers` is an admin-only API, so this means non-admins can't delete their own servers unlike `DELETE /users/{name}/servers/{server_name}` which is bad UX.

So I think our options are:

1. Continue with adding a dedicated `DELETE /servers/{id}` API which allows `@admin_or_self` and modify the `GET /user` and `GET /users/{name}` responses so they include the id of the user-owned servers.

    * Pros: admin tooling like `cull_idle_servers` would not need to get the user information when listing servers and deleting them making `GET /servers` faster (no extra join from `servers` to `users` in the DB).
    * Cons: more API changes and bloat (more than one way to delete a server), expanded maintenance cost

2. Scrap the idea of a dedicated `DELETE /servers/{id}` API and as long as `GET /servers` returns the user name the client can use the existing `DELETE /users/{name}/servers/{server_name}` API.

    * Pros: re-use existing API to delete a server that `cull_idle_servers` is already using, no new `DELETE /servers/{id}` API or changes to `GET /user` or `GET /users/{name}` to return the server id.
    * Cons: Slight performance hit on `GET /servers` to include the user name since we have to query that from the DB per server that is returned (joining the user when we query the servers might negate that, or we could cache the `user_id -> User` mapping if the same user owns more than one server).

In the case of issue #2954 the client is the `cull_idle_servers` script which runs with an admin token.

[1] https://github.com/jupyterhub/jupyterhub/pull/2967/files#diff-0a701bf780442df25926c4771a5f170eR30

Reply:
I've gone with option 2 and dropped pull #2967 once I got `cull_idle_servers` updated to use `GET /servers` in pull #2969.

Reply:
At least I use `state` in my hubs to adjust culling parameters within the cull_idle script.  It could be reasonable to be off by default, as long as it's obvious how to turn it on again, which I can add later if need be (but I guess it should be the same setting for /users and /servers?).

Reply:
> At least I use `state` in my hubs to adjust culling parameters within the cull_idle script. It could be reasonable to be off by default, as long as it's obvious how to turn it on again, which I can add later if need be

I would think that would be a query parameter if `state` is not returned by default from `GET /servers` so you could do something like:

`GET /servers?include_state=true`

Or something similar to get the `state` field returned if you need it.

> (but I guess it should be the same setting for /users and /servers?).

Yeah this is my main question and if so, it means we should only include `state` for active servers which is how the `GET /users` (`user_model`) code handles that. I'm fine either way, was just looking for direction from maintainers on this wrinkle.